pkg/kinflate/transformers/overlay.go

```diff
@@ -59,39 +65,143 @@ func (o *overlayTransformer) Transform(baseResourceMap resource.ResourceCollecti
 		switch {
 		case runtime.IsNotRegisteredError(err):
 			// Use JSON merge patch to handle types w/o schema
-			baseBytes, err := json.Marshal(base.Data)
+			merged, err = jsonMergePatch(base.Data.Object, overlay.Data.Object, jsonpatch.MergePatch)
 			if err != nil {
 				return err
 			}
-			patchBytes, err := json.Marshal(overlay.Data)
+		case err != nil:
+			return err
+		default:
+			// Use Strategic Merge Patch to handle types w/ schema
+			// TODO: Change this to use the new Merge package.
+			// Store the name of the base object, because this name may have been munged.
+			// Apply this name to the StrategicMergePatched object.
+			lookupPatchMeta, err := strategicpatch.NewPatchMetaFromStruct(versionedObj)
 			if err != nil {
 				return err
 			}
-			mergedBytes, err := jsonpatch.MergePatch(baseBytes, patchBytes)
+			merged, err = strategicpatch.StrategicMergeMapPatchUsingLookupPatchMeta(
+				base.Data.Object,
+				overlay.Data.Object,
+				lookupPatchMeta)
 			if err != nil {
 				return err
 			}
-			err = json.Unmarshal(mergedBytes, &merged)
+		}
+		base.Data.SetName(baseName)
+		baseResourceMap[gvkn].Data.Object = merged
+	}
+	return nil
+}
+
+// mergePatches merge and index patches by GVKN.
+// It errors out if there is conflict between patches.
+func (o *overlayTransformer) mergePatches() (resource.ResourceCollection, error) {
+	rc := resource.ResourceCollection{}
+	for _, patch := range o.overlay {
+		gvkn := patch.GVKN()
+		existing, found := rc[gvkn]
+		if !found {
+			rc[gvkn] = patch
+			continue
+		}
+		var merged map[string]interface{}
+		versionedObj, err := scheme.Scheme.New(gvkn.GVK)
+		switch {
```

Comment:
I have the feeling that this switch should be a "virtual table"/dispatch rather than a switch, this would make it much easier to test/understand.

Reply:
Not following.
Can you be more concrete?

Reply:
OK So let's look at the `case` statement. This one does:
- Check if there are conflicts,
- if there are, locate them and return,
- if we don't, merge the objects.

Now let's look at the default `case` statement. What do you do?
- Check if it has conflicts,
- if there are, locate them and return,
- if we don't, merge the objects.

That logic (check-locate-return-merge) is **orthogonal** to how each of these steps are actually performed (using jsonpath or smp). And that logic is reproduced twice.

I'd rather see something like this:
```
var mergeStrategy MergeStrategyInterface

// The following piece of code decides what type of strategy to use.
if !runtime.IsNotRegisteredError(err) {
   return err
}
if err != nil {
   mergeStrategy = JsonPatch{} // add other useful stuff in there
} else {
  mergeStrategy = SMP{Schema: schema} // add other useful stuff in there
}

// The following piece has the actual logic that we care about:
// If conflict, where are they, otherwise merge:
if mergeStrategy.HasConflict(object1, object2) {
   return mergeStrategy.FindConflict(object1, object2)
} else {
  return mergeStrategy.Merge(object1, object2)
}

...
// The details of how each strategy works come somewhere else...
```

The same applies for the two conflict detection functions below.

Reply:
I see. This is similar to what we do in apply v1.5 you developed.